tool/docs/cpp/rules.bzl

```diff
@@ -0,0 +1,129 @@
+
+def doxygen_to_adoc(name, data, docs_dirs):
+    args = ["$(location %s)" % target for target in data] + [
+        "--output",
+        "cpp/docs",
+    ] + ["--dir=%s=%s" % (filename, docs_dirs[filename]) for filename in docs_dirs]
+    kt_jvm_binary(
+        name = name,
+        srcs = [
+            "//tool/docs:cpp/DoxygenParser.kt",
+        ],
+        main_class = "com.vaticle.typedb.driver.tool.docs.cpp.DoxygenParserKt",
+        args = args,
+        deps = [
+            "//tool/docs:html_docs_to_adoc_lib",
+            "@maven//:org_jsoup_jsoup",
+            "@maven//:info_picocli_picocli",
+        ],
+        data = data,
+        visibility = ["//visibility:public"],
+    )
+
+def _doxygen_docs_impl(ctx):
+    output_directory = ctx.actions.declare_directory(ctx.attr._output_directory)
+    files = []
+    for target in ctx.attr.sources:
+        files.extend(target.files.to_list())
+
+    replacements = {
+        "PROJECT_NAME": ctx.attr.project_name,
+        "OUTPUT_DIRECTORY" : output_directory.path
+    }
+    if ctx.attr.strip_from_path != None:
+            replacements["STRIP_FROM_PATH"] =ctx.attr.strip_from_path
+
+    if ctx.file.main_page_md != None:
+        files.append(ctx.file.main_page_md)
+        replacements["USE_MDFILE_AS_MAINPAGE"] = ctx.file.main_page_md.path
+
+    replacements["INPUT"] = " ".join([f.path for f in files])
+
+    # Prepare doxyfile replacements
+    replacements_file = ctx.actions.declare_file("%s.replacements" % ctx.attr.name)
+    ctx.actions.write(
+        output = replacements_file,
+        content = "\n".join([(k + " = " +replacements[k]) for k in replacements]),
+        is_executable = False,
+    )
+
+    # Prepare doxyfile
+    doxyfile = ctx.actions.declare_file("%s.doxyfile" % ctx.attr.name)
+    ctx.actions.run(
+        inputs = [ctx.file._templater_script, ctx.file._doxyfile_template, replacements_file],
+        outputs = [doxyfile],
+        arguments = [ctx.file._templater_script.path, ctx.file._doxyfile_template.path, replacements_file.path, doxyfile.path],
+        executable = "python3",
+        use_default_shell_env = True
+    )
+    files = [doxyfile] + files
+    print(doxyfile.path)
+    ctx.actions.run(
+        inputs = files,
+        outputs = [output_directory],
+        arguments = [doxyfile.path],
+        executable = "doxygen",
+        use_default_shell_env = True
+    )
+
+    return DefaultInfo(files = depset([output_directory]))
+
+doxygen_docs = rule(
+    implementation = _doxygen_docs_impl,
+    test = False,
+    attrs = {
+        "project_name" : attr.string(
+            doc = "The project name for the doxygen docs",
+            mandatory = True,
+        ),
+        "sources" : attr.label_list(
+            doc = "A list of files made available to doxygen. This is NOT automatically included in the doxyfile",
+            mandatory = True,
+            allow_files = True,
+        ),
+        "strip_from_path" : attr.string(
```

Comment:
I think this is normally called 'strip_prefix' in bazel?

Reply:
Fair. It's called `strip_from_path` in doxygen.
It's the same with `main_page_md` and `project_name`.